Support for CRL

unreality opened the pull request:
### Description
This pull adds a new API endpoint `/crl` that will return an on-demand generated PEM-encoded CRL. Fixes #206 

The CRL is generated on-demand:
* if it is requested after the `NextUpdate` time specified in any previously generated CRL
* if a certificate has been revoked
* if no CRL has been generated previously

A new database table for storing the CRL named `x509_crl` is created when step-ca is launched.

A new interface `CertificateAuthorityCRLGenerator` is available for CAS object to implement. If they implement the interface, CRL lists will be created, otherwise a 'not implemented' error is raised when the request for CRL is made.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/smallstep/certificates?pullRequest=731) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/smallstep/certificates?pullRequest=731) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/smallstep/certificates?pullRequest=731) it.</sub>

maraino commented on authority/tls.go:
```diff
@@ -458,6 +464,95 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+
+	// Quick and dirty PEM encoding
+	// TODO: clean this up
+	pemCRL := fmt.Sprintf("-----BEGIN X509 CRL-----\n%s\n-----END X509 CRL-----\n", base64.StdEncoding.EncodeToString(certificateRevocationList))
```
```go
pem.EncodeToMemory(&pem.Block{
    Type: "X509 CRL",
    Bytes: certificateRevocationList,
})
```

unreality commented:
I changed `GenerateCertificateRevocationList` to return a DER formatted list, it seems many clients prefer to receive DER instead of PEM, so this should make it easier to support that.

J-Hunter-Hawke commented on authority/tls.go:
```diff
@@ -458,6 +463,91 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
 	return a.db.Revoke(rci)
 }
 
+// GenerateCertificateRevocationList returns a PEM representation of a signed CRL.
```
I'm splitting hairs a little, but the function is using DER encoding

unreality replied:
Whoops, missed some mentions of PEM when i switched it over to DER format :) Thanks

maraino commented:
@unreality You're right about the DER format, if you look at the CRL distributions points embedded in certificates, they are DER encoded.

I will talk about this PR with the team in our triage meeting, I have some ideas that want to bring to the team before moving forward.

maraino commented on cas/softcas/softcas.go:
```diff
@@ -112,6 +113,17 @@ func (c *SoftCAS) RevokeCertificate(req *apiv1.RevokeCertificateRequest) (*apiv1
 }
 
+// CreateCertificateRevocationList will create a new CRL based on the RevocationList passed to it
+func (c *SoftCAS) CreateCertificateRevocationList(crl *x509.RevocationList) ([]byte, error) {
```
I think we want to make it more extensible and add a `*apiv1.CreateCertificateRevocationListRequest` and `*apiv1.CreateCertificateRevocationListResponse`, although initially they will only have the CRL on the request and return a field named CRL too (or SignedCRL, ASN1Bytes, what do you think is better).

maraino replied:
Perhaps we can name everything `CreateCRL*`, `CreateCertificateRevocationListRequest` seems too long.

maraino commented on authority/tls.go:
```diff
@@ -426,6 +428,9 @@ func (a *Authority) Revoke(ctx context.Context, revokeOpts *RevokeOptions) error
 		err = a.revoke(revokedCert, rci)
+
+		// Generate a new CRL so CRL requesters will always get an up-to-date CRL whenever they request it
+		_, _ = a.GenerateCertificateRevocationList(true)
```
I'd like you to change the way of generating CRLs to a background job, and every time a certificate is revoked, a message will be sent to a channel, that channel should probably be waiting on a `select` until a new message is sent or until a timer expires:
```go
select {
case <- a.revokeCh:
	// GenerateCRL
case <-time.After(a.revokeDuration): // it might be good to add some jitter here.
	// GenerateCRL
}

maraino replied:
With this job, we should rename the method `GenerateCertificateRevocationList` to something like `GetCertificateRevocationList` as it will be always like a static file.

maraino commented on authority/tls.go:
```diff
@@ -458,6 +463,91 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+		RevokedCertificates: revokedCertificates,
+		Number:              &bn,
+		ThisUpdate:          time.Now().UTC(),
+		NextUpdate:          time.Now().UTC().Add(time.Minute * 10),
```
Making the value configuration we might be able to disable (0m) on a "non-master" node if the CA is running distributed using a shared MySQL database.

maraino replied:
In fact, it might be good to just enable the CRL if configured:
```
// Enabled and generate CRL every 8h
"crl": {"enabled": true, "cacheDuration": "8h"}
// Enabled but do not generate the CRL
"crl": {"enabled": true}
"crl": {"enabled": true, "cacheDuration": "0h"}
// Disabled
// Enabled but do not generate the CRL
"crl": {"enabled": false}
```

maraino replied:
So this method should be split in two, one like `GetCertificateRevocationList` that returns the CRL from the DB if enabled, and return a 404 or 501 error if not.

And a background job that will start only if enabled and a `cacheDuration` is greater than 0. This job will also be notified with a channel after a revocation.

unreality replied:
I changed the config slightly to (hopefully) make the behaviour more clear:
`"crl": {"generate": true, "cacheDuration": "8h"}`

`generate` controls whether the automatic generation goroutine is created
`cacheDuration` is also needed on all nodes (if using multiple nodes) so that when a certificate is revoked and the node creates a new CRL, the node has a duration to place in the CRL.

if the `crl` dict is missing, CRL generation/serving is completely disabled.

maraino commented on api/crl.go:
```diff
@@ -0,0 +1,24 @@
+	})
+
+	w.WriteHeader(200)
+	_, err = w.Write(pemBytes)
```
To avoid the linter error you can do `fmt.Fprint(w, pemBytes)`

maraino commented on api/crl.go:
```diff
@@ -0,0 +1,24 @@
+		return
+	}
+
+	pemBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "X509 CRL",
+		Bytes: crlBytes,
+	})
```
We should use the DER format. At least by default/

unreality replied:
I added a http param to select the format, eg:

`https://ca.example.com/crl` returns `DER`
`https://ca.example.com/crl?pem` returns `PEM`

maraino reviewed (changes requested):
Hi @unreality, this is great. But after discussing it with my team, we want a few changes, mainly:
* Convert the generation to a job (background goroutine)
* Make it configurable
* Do not include expired certificates

unreality commented on authority/tls.go:
```diff
@@ -458,6 +463,91 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+
+	// some CAS may not implement the CRLGenerator interface, so check before we proceed
+	caCRLGenerator, ok := a.x509CAService.(casapi.CertificateAuthorityCRLGenerator)
+
```
Im unsure what you mean by 'white lines' ? Is it that the comment gives no additional information, or the whitespace is not how you want it formatted? Or something else? :)

maraino commented:
@unreality I need to look at this again, but can you do some experiments and see how the enumeration of revokes performs with millions of revokes?

unreality commented:
I'll see if i can test the speed of the crl gen based on a large number of revoked certs, but no promises :)

I'll also try fix all the linter errors im seeing.

Another issue ive come across is that when attempting to verify the CRL under windows, the mechanism complains about needing a client certificate to get the crl:

`certutil -URL https://yourca.example.com/crl` 

Returns:

`Error retrieving URL: A certificate is required to complete client authentication 0x80072f0c (WinHttp: 12044 ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED)`

I'm assuming this is because step-ca prompts for mTLS on any api endpoint? Is there any way of excluding a route from prompting for mTLS?

maraino commented:
> I'm assuming this is because step-ca prompts for mTLS on any API endpoint? Is there any way of excluding a route from prompting for mTLS?

`step-ca` only requires a client certificate (mTLS) in the /renew endpoint. If you send it to any other endpoint it will be verified, but it's not required. We basically use Go's `tls.VerifyClientCertIfGiven` and then on the methods that require mTLS we check if a certificate was given.

My guess is that that certutil might have a bug in how it manages client certificates, but you can always create a client certificate. By default all certificates signed by `step-ca` can be used for client authentication. This is the base on how or `/renew` endpoint works.

maraino commented on authority/tls.go:
```diff
@@ -458,6 +476,118 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+// error if the underlying AuthDB does not support CRLs
+func (a *Authority) GetCertificateRevocationList() ([]byte, error) {
+	if a.config.CRL == nil {
+		return nil, errs.Wrap(http.StatusInternalServerError, errors.Errorf("Certificate Revocation Lists are not enabled"), "authority.GetCertificateRevocationList")
```
We can return `http.StatusNotFound` if it's not enabled. Or perhaps `http.StatusNotImplemented`. Not sure which makes more sense.

maraino reviewed (changes requested):
I've added some comments, the tests or linters are also failing can you fix the errors.

dopey commented:
Hey @unreality let us know if you have any questions about the feedback. Cheers 🍻 and Happy Holidays!

unreality commented:
Apologies for the delay on this one, now im back at work i hope i'll be able to get this cleaned up for merging soon

LarsBingBong commented:
Will this also make OCSP work? Me asking in > https://discord.com/channels/837031272227930163/841243933936975912/939177692874555502

maraino commented:
@LarsBingBong an OSCP responder can be created combining a couple of things, see details 
https://gist.github.com/maraino/2905ba742168d04ddca9882ac1f0c3ed and here for badger https://gist.github.com/maraino/4dcb64cb051b17ef6d421892cb4e55a8

unreality commented on authority/tls.go:
```diff
@@ -362,6 +364,16 @@ func (a *Authority) Revoke(ctx context.Context, revokeOpts *RevokeOptions) error
 		err error
 	)
+
+	// Attempt to get the certificate expiry using the serial number.
+	cert, err := a.db.GetCertificate(revokeOpts.Serial)
+
+	// Revocation of a certificate not in the database may be requested, so fill in the expiry only
+	// if we can
+	if err == nil {
+		rci.ExpiresAt = cert.NotAfter
+	}
```
Where is the info for the mTLS auth stored? Its been a while since looking at this, so it may be quicker for you to point me at the NotAfter that can be used in mTLS calls... :)

maraino replied:
RevokeOptions has a certificate if mTLS is used. See 
https://github.com/smallstep/certificates/blob/5cb23c6029fe775a303b5613f6d4374c4df2550c/authority/tls.go#L363-L372

A way to test this is to generate a certificate and then revoke it using: 
```console
step ca revoke --cert foo.crt --key foo.key
```

Instead of using just the serial.

maraino replied:
Basically, this lookup is only necessary if Crt is nil.

unreality commented:
I've pushed most of the requested changes, let me know if there are still issues that need resolving :)

maraino commented on cas/softcas/softcas.go:
```diff
@@ -112,6 +113,17 @@ func (c *SoftCAS) RevokeCertificate(req *apiv1.RevokeCertificateRequest) (*apiv1
+// CreateCRL will create a new CRL based on the RevocationList passed to it
+func (c *SoftCAS) CreateCRL(req *apiv1.CreateCRLRequest) (*apiv1.CreateCRLResponse, error) {
+
+	revocationListBytes, err := x509.CreateRevocationList(rand.Reader, req.RevocationList, c.CertificateChain[0], c.Signer)
```
The signer and certificate chain should not be used here. The last version in the `master` branch has a method that returns the right one:
https://github.com/smallstep/certificates/blob/abf5fc32a3417ca95800bf1d4c8cef2030cf8782/cas/softcas/softcas.go#L210-L216

maraino commented on db/db.go:
```diff
@@ -188,6 +208,73 @@ func (db *DB) RevokeSSH(rci *RevokedCertificateInfo) error {
+			return nil, err
+		}
+
+		if !data.ExpiresAt.IsZero() && data.ExpiresAt.After(time.Now().UTC()) {
```
Before the for loop, define a fixed timestamp and use that every time.

```golang
now := time.Now().Truncate(time.Second)
```

I would also truncate it to the second, and there's no need to use UTC, After will take care of that.

maraino commented on db/db.go:
```diff
@@ -188,6 +208,73 @@ func (db *DB) RevokeSSH(rci *RevokedCertificateInfo) error {
+				continue
+			}
+
+			if cert.NotAfter.After(time.Now().UTC()) {
```
Use the fixed time from my previous comment.

maraino commented on db/db.go:
```diff
@@ -24,6 +25,9 @@ var (
 	sshHostPrincipalsTable = []byte("ssh_host_principals")
 )
 
+var crlKey = []byte("crl") //TODO: at the moment we store a single CRL in the database, in a dedicated table.
+//      is this acceptable? probably not....
+
```
Can you use the "AuthorityID" as the key? This id is not generally used in all scenarios, but some of them are using it. The id can be read from `a.config.AuthorityConfig.AuthorityID` but if it's empty you can use the default one:

```golang
const DefaultAuthorityID = "00000000-0000-0000-0000-000000000000"
```

maraino commented on db/db.go:
```diff
@@ -188,6 +208,73 @@ func (db *DB) RevokeSSH(rci *RevokedCertificateInfo) error {
+}
+
+// GetCRL gets the existing CRL from the database
+func (db *DB) GetCRL() (*CertificateRevocationListInfo, error) {
```
This will require the authority id I mention before.

maraino commented on db/db.go:
```diff
@@ -188,6 +208,73 @@ func (db *DB) RevokeSSH(rci *RevokedCertificateInfo) error {
+}
+
+// StoreCRL stores a CRL in the DB
+func (db *DB) StoreCRL(crlInfo *CertificateRevocationListInfo) error {
```
The authority id can also be a field in the DB type, as it won't change. And the interface will remain as it is.

unreality replied:
im unsure how DB can access the configured AuthorityID without altering either the DB config, or the interface? this could be because i am looking at an older branch, but DB does not currently have access to the AuthorityConfig as far as i can see...?

maraino commented on api/crl.go:
```diff
@@ -0,0 +1,52 @@
+		return
+	}
+
+	if crlBytes == nil {
+		w.WriteHeader(404)
+		_, err = fmt.Fprintln(w, "No CRL available")
+		if err != nil {
+			panic(errors.Wrap(err, "error writing http response"))
+		}
+		return
+	}
```
Use [`http.NotFound(w, r)`](https://pkg.go.dev/net/http#NotFound) here

maraino replied:
Well in fact, `GetCertificateRevocationList` returns a 404 if CRLs are not enabled, we can detect that error and use `http.NotFound`.

unreality replied:
I used the generic http.Error() instead so its a bit neater

maraino commented on api/api.go:
```diff
@@ -46,6 +46,8 @@ type Authority interface {
 	GetRoots() (federation []*x509.Certificate, err error)
 	GetFederation() ([]*x509.Certificate, error)
 	Version() authority.Version
+	GenerateCertificateRevocationList() error
```
GenerateCertificateRevocationList is not used in the api package, it should not be here.

maraino commented on api/api_test.go:
```diff
@@ -580,6 +580,10 @@ type mockAuthority struct {
 	version                      func() authority.Version
 }
 
+func (m *mockAuthority) GenerateCertificateRevocationList(force bool) ([]byte, error) {
+	panic("implement me")
+}
+
```
There's no need for this.

maraino replied:
But you might need `GetCertificateRevocationList`.

maraino commented on authority/authority.go:
```diff
@@ -615,3 +629,46 @@ func (a *Authority) requiresSCEPService() bool {
+				err := a.GenerateCertificateRevocationList()
+				if err != nil {
+					log.Printf("ERROR: authority.crlGenerator encountered an error when regenerating the CRL: %v", err)
+				}
```
We should this loop and return in `Shutdown()` and `CloseForReload()`, in fact as it is right now, an extra goroutine will be running each time we reload step-ca using the HUP signal.

maraino commented on authority/authority.go:
```diff
@@ -615,3 +629,46 @@ func (a *Authority) requiresSCEPService() bool {
+
+func (a *Authority) startCRLGenerator() error {
+
+	if a.config.CRL.CacheDuration.Duration <= 0 {
```
You're already doing this check above, you can move the if conditions from above here.

LarsBingBong commented:
> @LarsBingBong an OSCP responder can be created combining a couple of things, see details https://gist.github.com/maraino/2905ba742168d04ddca9882ac1f0c3ed and here for badger https://gist.github.com/maraino/4dcb64cb051b17ef6d421892cb4e55a8

@maraino I can't see how I can use this together with the Helm chart to the step-ca. There's no `go` client/cli installed. So it can't be executed inside the `step-ca` container on Kubernetes. And the PVC is `RWO` so it can't be mounted to multiple `POD's`.

maraino commented:
> > @LarsBingBong an OSCP responder can be created combining a couple of things, see details https://gist.github.com/maraino/2905ba742168d04ddca9882ac1f0c3ed and here for badger https://gist.github.com/maraino/4dcb64cb051b17ef6d421892cb4e55a8
> 
> @maraino I can't see how I can use this together with the Helm chart to the step-ca. There's no `go` client/cli installed. So it can't be executed inside the `step-ca` container on Kubernetes. And the PVC is `RWO` so it can't be mounted to multiple `POD's`.

Those scripts are not supported at all, badger only supports one reader at the time, it might work if you copy the database files to a new directory, but you will also need a new image with the required scripts and the binary version of that .go.

And with this PR merged, you will be also able to create an OCSP responder using the generated CRL as the source.

unreality commented:
Committed some additional changes, but i'll need to merge or rebase to continue i think, so i'll do that when I have some time

LarsBingBong commented:
> > > @LarsBingBong an OSCP responder can be created combining a couple of things, see details https://gist.github.com/maraino/2905ba742168d04ddca9882ac1f0c3ed and here for badger https://gist.github.com/maraino/4dcb64cb051b17ef6d421892cb4e55a8
> > 
> > 
> > @maraino I can't see how I can use this together with the Helm chart to the step-ca. There's no `go` client/cli installed. So it can't be executed inside the `step-ca` container on Kubernetes. And the PVC is `RWO` so it can't be mounted to multiple `POD's`.
> 
> Those scripts are not supported at all, badger only supports one reader at the time, it might work if you copy the database files to a new directory, but you will also need a new image with the required scripts and the binary version of that .go.
> 
> And with this PR merged, you will be also able to create an OCSP responder using the generated CRL as the source.

Yeah that was my conclusion as well. Fully confirmed with your comments. Much appreciated. Thanks! Looking forward to this merging at some point - in hopefully the near future.

logopk commented:
> > > @LarsBingBong an OSCP responder can be created combining a couple of things, see details https://gist.github.com/maraino/2905ba742168d04ddca9882ac1f0c3ed and here for badger https://gist.github.com/maraino/4dcb64cb051b17ef6d421892cb4e55a8
> > 
> > 
> > @maraino I can't see how I can use this together with the Helm chart to the step-ca. There's no `go` client/cli installed. So it can't be executed inside the `step-ca` container on Kubernetes. And the PVC is `RWO` so it can't be mounted to multiple `POD's`.
> 
> Those scripts are not supported at all, badger only supports one reader at the time, it might work if you copy the database files to a new directory, but you will also need a new image with the required scripts and the binary version of that .go.
> 
> And with this PR merged, you will be also able to create an OCSP responder using the generated CRL as the source.

@maraino could you please explain how this can be done?

Currently I'm using openssl ocsp, that needs the index.txt to serve the ocsp responses. That's why I created the listcerts.go for badger to create the plaintext file.

Do you know of any good ocsp responder project that can rely on crls(deltas preferred)?

Thanks.

philfry commented:
Hi all,

> Do you know of any good ocsp responder project that can rely on crls(deltas preferred)?

Correct me if I'm wrong, but an ocsp responder needs to know about revoked certificates as well as about valid ones to provide a valid answer.

logopk commented:
> Hi all,
> 
> > Do you know of any good ocsp responder project that can rely on crls(deltas preferred)?
> 
> Correct me if I'm wrong, but an ocsp responder needs to know about revoked certificates as well as about valid ones to provide a valid answer.

That would be the point of openssl ocsp relying on the (complete) index.txt.

maraino commented:
@philfry @logopk Yes, an OCSP responder requires to know about the active certificates. If your only source of information is the CRL you will need to fake the response of any other certificate and return the status GOOD for any other serial number.

And because you need to fake responses there won't be any OCSP responder available. You'll need to build it yourself, it's not hard.

unreality commented:
ive completed all the changes except for the AuthorityID stuff, since i couldnt find a good way of accessing it. also its probably a mess now that i've rebased it :/

LarsBingBong commented:
Good stuff @unreality - thank you!

@maraino - when this gets merged. If I'm consuming the step-ca via a Helm chart on Kubernetes. [Step Certificates Helm chart](https://github.com/smallstep/helm-charts/tree/master/step-certificates) will also need to be updated right?

Thanks

logopk commented:
> @philfry @logopk Yes, an OCSP responder requires to know about the active certificates. If your only source of information is the CRL you will need to fake the response of any other certificate and return the status GOOD for any other serial number.
> 
> And because you need to fake responses there won't be any OCSP responder available. You'll need to build it yourself, it's not hard.

@maraino OK, I absolutely agree. I'm sorry but you got me confused with the comment:
> And with this PR merged, you will be also able to create an OCSP responder using the generated CRL as the source.
That's why I asked. Maybe you can explain this in detail?

Thank you.

larssb commented:
Almost there - any day now 👍  .... the hold up is the minor stuff in this PR right? @unreality do you kindly know?

LarsBingBong commented:
Any news here @unreality and or @maraino - thank you very much 👍🏿

unreality commented:
Ive brought this branch up to latest master, @maraino could you review and let me know any changes that are needed?

LarsBingBong commented:
@maraino a polite request for a PR review 👍🏿 - are we there now?

Thanks

LecrisUT commented:
@unreality A few things:
- Could you add appropriate flags to `step ca init` to enable this in the `ca.json` generation
- Clarify and/or add the options to add the CRL to either or both the root and intermediate certificates. For example there could be a school root certificate with its own CRL to invalidate intermediates, and lab intermediate CRL to invalidate their own managed services
- A few simple words in the [docs section](https://github.com/smallstep/docs/blob/main/src/pages/docs/step-ca/configuration.mdx) to make it known

maraino commented:
@unreality @LecrisUT I think I will have time this week to look into this. But there's one thing to consider: we don't see this as the final implementation, and it might go away at some point once authority webhooks are baked into step-ca.

Those webhooks will act on different certificate events. For example, when a certificate is issued, you will be able to configure a webhook that will send the issued certificate (or perhaps a pre-certificate) to a third party, so it can store it or perhaps, in some cases, perform some validation, the exact details are still not designed. In the same way, when a certificate is revoked, we will send the serial and the certificate information we have to a third party that can use this to generate the CRLs, and store them in a bucket, so it's accessible to the users.

If we implement any of those third-party services is not yet decided, but the revocation one is probably the best candidate.

LecrisUT commented:
By third party here do you mean a golang library that already codes the CRL part, or a separately service from `step-ca`? What would be the benefit of having it managed by a third party over it being served by `step-ca`?

maraino commented:
> By third party here do you mean a golang library that already codes the CRL part, or a separately service from step-ca? What would be the benefit of having it managed by a third party over it being served by step-ca?

In a Golang binary, for example, the main benefit would be avoiding having `step-ca` in a more public network than necessary, in case of clients need to check CRLs or OCSPs. That binary might, for example, use `gocloud.dev` to store the new CRL using a cloud storage service, as well as properly removing expired certs.

LecrisUT commented:
Hmm good point. I would personally handle those via reverse proxy, serving only the crl/ocsp endpoints to a public url (also allowing to use a publicly signed certificate). I think there is still some merit to implementing these features locally, if nothing else, in order to make managing easier.

logopk commented:
At least will we need the CRL on a non https-Port. Which would separate it from the CA-Port.

LecrisUT commented:
> At least will we need the CRL on a non https-Port

If the goal is to keep it simple, that is not a requirement, although it would be good to have. For example `caddy` can reverse proxy to an https port, which by default checks the integrity of the connection, strips the original TLS, and serves its own. But whether or not there is an RFC requirement about how to server crl/ocsp endpoints, that might dictate how to serve/reverse-proxy these.

logopk commented:
CRL by definition requires non-TLS.

maraino commented:
Currently, only SCEP can use HTTP if `insecureAddress` is set. We can consider adding CRLs to on a separate PR.

LarsBingBong commented:
I see it's now on the v0.23.0 milestone. Looking at the release cycle of this project, is it then fair to say we're looking at a couple of months before this is out?

Thank you very much

unreality commented:
I've updated this to master to resolve conflicts.

@maraino what needs to be done to move this forward, I see @areed is doing a lot of work on webhooks, are these the authority webhooks you mention? Perhaps I should re-write the CRL support to use those instead?

maraino commented:
> ... I should re-write the CRL support to use those instead?

I don't think all the necessary things will be ready. 

I'll be looking at this pretty soon, I was planning on resolving those conflicts myself.

maraino commented on db/db.go:
```diff
@@ -220,6 +240,75 @@ func (db *DB) RevokeSSH(rci *RevokedCertificateInfo) error {
+		}
+
+		if !data.RevokedAt.IsZero() && data.RevokedAt.After(now) {
+			revokedCerts = append(revokedCerts, data)
```
There won't be certificates revoked after now. While testing this I've only seen the certificate I've just revoked because this method is called just after the revocation, and with the truncation of now, this condition might be true.

maraino replied:
The CRL must contain all certificates that are still valid or will be valid but have been revoked.

unreality replied:
I removed the `After()` check, im unsure why it was there, i was probably confused about the date calculation? Let me know if its the right calc now, im not confident i havent done something fuzzy :)

maraino commented on authority/config/config.go:
```diff
@@ -183,6 +192,9 @@ func (c *Config) Init() {
 	}
+	if c.CRL != nil && c.CRL.Generate && c.CRL.CacheDuration == nil {
+		c.CRL.CacheDuration = DefaultCRLCacheDuration
+	}
```
I added this to your PR

maraino commented on authority/authority.go:
```diff
@@ -633,6 +637,16 @@ func (a *Authority) init() error {
 
+	// Start the CRL generator
+	if a.config.CRL != nil && a.config.CRL.Generate {
+		if v := a.config.CRL.CacheDuration; v != nil && v.Duration > 0 {
```
I've fixed a panic here.

maraino commented on api/crl.go:
```diff
@@ -0,0 +1,32 @@
+	crlBytes, err := mustAuthority(r.Context()).GetCertificateRevocationList()
+	if err != nil {
+		render.Error(w, err)
+		return
```
I added this to your PR

maraino commented on authority/config/config.go:
```diff
@@ -183,6 +192,9 @@ func (c *Config) Init() {
 	if c.CommonName == "" {
 		c.CommonName = "Step Online CA"
 	}
+	if c.CRL != nil && c.CRL.CacheDuration == nil {
+		c.CRL.CacheDuration = DefaultCRLCacheDuration
+	}
```
I added this to your PR to fix a panic in `GenerateCertificateRevocationList`

maraino commented on authority/config/config.go:
```diff
@@ -106,6 +109,12 @@ type AuthConfig struct {
 	DisableGetSSHHosts   bool                  `json:"disableGetSSHHosts,omitempty"`
 }
 
+// CRLConfig represents config options for CRL generation
+type CRLConfig struct {
+	Generate      bool                  `json:"generate,omitempty"`
+	CacheDuration *provisioner.Duration `json:"cacheDuration,omitempty"`
+}
```
The configuration here is weird, First, I fixed a panic with `"crl": {}`, on revocation there was a panic on `GenerateCertificateRevocationList` because  `CacheDuration` was nil. Then if generate is false, the CRL won't be renewed unless a certificate is revoked. So you will be able to get a CRL, but it will be expired.

unreality replied:
This weirdness is to support multiple nodes:
- One node should have `generate: true` so that a CRL gets generated regularly
- Additional nodes should have `cacheDuration: XXm` and `generate: false` so that if a revoke request is routed to the extra nodes, it can successfully generate a CRL and place it in the DB with the correct duration set

I'll add some logic to test for 0-length durations though, unless you would like the config refactored? It is a bit weird, so im open to suggestions

maraino replied:
The second node doesn't need to create a CRL to store the proper `ExpiresAt` in the DB, what feels weird is that it will generate a CRL that will expire and won't be regenerated unless another revoke is routed to that node.

What makes sense to me is having one node that generates every XX time (24h by default) a new CRL, and perhaps optionally regenerates it if a new revoke occurs. Although this is not common as clients can have in their cache the old CRL because it hasn't expired.

unreality replied:
I'm a little confused, perhaps i've misunderstood something somewhere.

The multiple nodes scenario i'm looking at is the following:
* `Node A` connected to `SharedDB` sitting behind `LoadBalancer`
* `Node B` connected to `SharedDB` sitting behind `LoadBalancer`

Lets assume `Node A` has `crl: {"generate": true, "cacheDuration": "5m"}`

`Node A` starts up and generates a CRL regularly, storing it in `SharedDB`. `ClientA` requests the CRL regularly from the endpoint, the request being routed randomly between `Node A` and `Node B`.

`Client B` sends a revoke request, and it is routed the `Node B`. `Node B` will add the revoked cert to the list of revoked certificates in `SharedDB` and initiate regeneration of the CRL, and in doing so needs to place a value into the `NextUpdate` field in the CRL. Where should `Node B` get the duration for calculating `NextUpdate` from? At present `Node B` uses `cacheDuration` to calculate it. Perhaps it could be calculated from the old CRL instead?

In this scenario im assuming we only want one node to be regularly generating CRLs, so one node has `generate: true` while all others have it set to `false`

I think there is also a concurrency issue where `Node A` could be regenerating the CRL while `Node B` is handling a revoke request, but im presuming that could be handled by the database?

Of course its entirely possible I have misunderstood your point here, so let me know if thats the case :)

maraino commented on authority/tls.go:
```diff
@@ -569,6 +592,118 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+
+// GenerateCertificateRevocationList generates a DER representation of a signed CRL and stores it in the
+// database. Returns nil if CRL generation has been disabled in the config
+func (a *Authority) GenerateCertificateRevocationList() error {
```
We should not run this concurrently. Right now, it will be executed n times if multiple revokes occur simultaneously; they might also collide with the automatic generation.

If we want to simplify this I think it's ok for now to have the automatic generation. New revocations won't appear until the next CRL is generated.

unreality replied:
Ive added a simple mutex so its no longer concurrent, i think that should be sufficient?

maraino commented on authority/tls.go:
```diff
@@ -569,6 +592,118 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
+		RevokedCertificates: revokedCertificates,
+		Number:              &bn,
+		ThisUpdate:          time.Now().UTC(),
+		NextUpdate:          time.Now().UTC().Add(a.config.CRL.CacheDuration.Duration),
```
I've fixed a panic here.

maraino reviewed (changes requested):
CRLs do not contain the expected certificates.

LarsBingBong commented:
Nice nice nice ... we're inching closer on this one .... wow!

unreality commented:
@maraino are you able to comment on the config format changes, mutix lock etc? I'm happy to alter things if the pull request still need some work, just trying to close this off so i can work on some other things :)

maraino commented:
@unreality What about this:

```go
type CRLConfig struct {
	Enabled          bool                  `json:"enabled"`
	CacheDuration    *provisioner.Duration `json:"cacheDuration,omitempty"`
	GenerateOnRevoke bool                  `json:"generateOnRevoke,omitempty"`
	RenewPeriod      *provisioner.Duration `json:"renewPeriod,omitempty"`
}
```

* The CRL will be available only if `crl.enabled` is true. 
* The default cache duration would be 24h unless `crl.cacheDuration` is set. If set, it must be > 0, although we can consider 0 as the default of 24h. The regeneration should happen before the expiration occurs. 
* `crl.generateOnRevoke` won't wait until the next generation, it will regenerate the CRL when a revoke occurs. It can either use the previous not-after or reset the regeneration timer appropriately.
* Optionally we can a `crl.renewBefore` or `crl.renewPeriod` to manage how the timer works. For example `cacheDuration` can be 24h but `renewPeriod` can be 8h. By default, it can be ~2/3 of the duration

I'm not sure if `generateOnRevoke` is a good name, others can be considered, `liveCRL` but the concept is the same.

unreality commented:
@maraino I have implemented the changes to the config, and added some tests that seem to be failing on CI for some reason I cant determine

Let me know if further changes are required

plinss commented:
This is a great feature and I'm reluctant to ask for anything that will delay this landing (FWIW, I'm already using it in a production system), but it would be helpful if the CRL endpoint returned a 'Last-Modified' header and respected the 'If-Modified-Since' request header.

The 'Last-Modified' should return the date the CRL was last regenerated, and requests with a 'If-Modified-Since' header would get an empty response with a 304 status code if the CRL wasn't regenerated since the time in the 'If-Modified-Since' header.

My use case is that Nginx requires a current CRL to be loaded and maintained by an external service. When the CRL is updated, Nginx needs to be reloaded to read it from disk. Given a cronjob that periodically fetches the current CRL, this allows the cronjob to know if Nginx needs to be reloaded or not, avoiding unnecessary reloads (which flushes other cached data, like OCSP staples).

larssb commented:
Curious on where we stand on this one ... are we still inching closer and about to hit this feature to be released in v0.23.0?

Thank you very much.

maraino commented:
> Curious on where we stand on this one ... are we still inching closer and about to hit this feature to be released in v0.23.0?

I'm not sure if all the issues that I mentioned were addressed. but I need to check again.

maraino commented:
Hi @unreality, I've made some changes to your branch, see 8200d19. If you're ok with those, I will proceed to resolve the conflicts and merge this PR.

maraino reviewed (commented):
Hi @unreality, I've made some changes to your branch, see https://github.com/smallstep/certificates/commit/8200d198941b370cc197e8c06075085c1c97c08b. If you're ok with those, I will proceed to resolve the conflicts and merge this PR.

unreality commented:
Hi @maraino, it looks fine to me, thanks for taking a look

maraino reviewed (approved):
🎺🥁🎉💥

LarsBingBong commented:
Can't wait for v0.23 to hit the roads.
